Source/Dafny/Verifier/Translator.cs

```diff
@@ -6260,7 +6260,37 @@ private class WFOptions {
 
       } else if (expr is StmtExpr) {
         var e = (StmtExpr)expr;
+        // If we're inside an "old" expression, then "etran" will know how to translate
```

Comment:
Can you help me understand better why we need to temporarily clear the heap here?

Reply:
A Boogie procedure operates on global variables. Dafny uses just one global variable, namely `$Heap`. (Actually, there's another one, too, called `$Tick`, but I don't think it's used anymore or could be deleted.) If a Dafny call statement sits inside an `old`, we would want the precondition of the Boogie procedure to use `old($Heap)` instead of `$Heap`. A Boogie procedure has no facility to parameterize the global variables it uses. Therefore, this PR changes the global variable `$Heap` to the desired value before any such Boogie procedure call is made.

This would not be a good idea if the procedure could change the heap. In fact, if that could happen, then we shouldn't allow the Dafny code to make such a call inside `old`. But lemma don't have any net effect on `$Heap`, so we're okay. After the calls, this PR thus changes the global variable `$Heap` back to what it previously had been.

So, essentially, the temporary assignment to `$Heap` is a way to pass in a different value for `$Heap` to any procedure calls that occur in the `Statement` part of the `StmtExpr`.

Reply:
Sorry, I should have asked a more specific question. I would expect us to set the global heap variable to some value that reflects the old heap, but it seems we're just clearing the heap. Am I reading this incorrectly, or how is _clearing_ the heap sufficient?

Reply:
This PR generates:

```
tmpHeap := $Heap;
$Heap := etran.$Heap;
...
$Heap := tmpHeap;
```

The RHS of the first assignment to `$Heap` (which I sketched as `etran.$Heap` here and in the code) is whatever the current `etran` thinks the `.HeapExpr` is. For example, this `etran.HeapExpr` may be `old($Heap)` or `$SpecialHeapForSomeSpecialPurpose`.

So, the heap is not empty, but is what the expression translator thinks the heap's value is.

Reply:
I see, sorry I think I should have just read better :)